plugins/bkpr/bookkeeper.c

```diff
@@ -73,12 +74,297 @@ static struct command_result *json_list_balances(struct command *cmd,
+			maybe_update_account(db, acct, chain_ev, tags);
+			db_commit_transaction(db);
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/* Net out credit/debit --> basically find the diff */
+static char *msat_net(const tal_t *ctx,
+		      struct amount_msat credit,
+		      struct amount_msat debit,
+		      struct amount_msat *credit_net,
+		      struct amount_msat *debit_net)
+{
+	if (amount_msat_eq(credit, debit)) {
+		*credit_net = AMOUNT_MSAT(0);
+		*debit_net = AMOUNT_MSAT(0);
+	} else if (amount_msat_greater(credit, debit)) {
+		if (!amount_msat_sub(credit_net, credit, debit))
+			return tal_fmt(ctx, "unexpected fail, can't sub."
+				       " %s - %s",
+				       type_to_string(ctx, struct amount_msat,
+						      &credit),
+				       type_to_string(ctx, struct amount_msat,
+						      &debit));
+		*debit_net = AMOUNT_MSAT(0);
+	} else {
+		if (!amount_msat_sub(debit_net, debit, credit)) {
+			return tal_fmt(ctx, "unexpected fail, can't sub."
+				       " %s - %s",
+				       type_to_string(ctx,
+					       struct amount_msat,
+					       &debit),
+				       type_to_string(ctx,
+					       struct amount_msat,
+					       &credit));
+		}
+		*credit_net = AMOUNT_MSAT(0);
+	}
+
+	return NULL;
+}
+
+static char *msat_find_diff(struct amount_msat balance,
+			    struct amount_msat credits,
+			    struct amount_msat debits,
+			    struct amount_msat *credit_diff,
+			    struct amount_msat *debit_diff)
+{
+	struct amount_msat net_credit, net_debit;
+	char *err;
+
+	err = msat_net(tmpctx, credits, debits,
+		       &net_credit, &net_debit);
+	if (err)
+		return err;
+
+	/* If we're not missing events, debits == 0 */
+	if (!amount_msat_zero(net_debit)) {
+		assert(amount_msat_zero(net_credit));
+		if (!amount_msat_add(credit_diff, net_debit, balance))
+			return "Overflow finding credit_diff";
+		*debit_diff = AMOUNT_MSAT(0);
+	} else {
+		assert(amount_msat_zero(net_debit));
+		if (amount_msat_greater(net_credit, balance)) {
+			if (!amount_msat_sub(debit_diff, net_credit,
+					    balance))
+				return "Err net_credit - amt";
+			*credit_diff = AMOUNT_MSAT(0);
+		} else {
+			if (!amount_msat_sub(credit_diff, balance,
+					     net_credit))
+				return "Err amt - net_credit";
+
+			*debit_diff = AMOUNT_MSAT(0);
+		}
+	}
+
+	return NULL;
+}
+
+static void log_journal_entry(struct account *acct,
+			      const char *currency,
+			      u64 timestamp,
+			      struct amount_msat credit_diff,
+			      struct amount_msat debit_diff)
+{
+	struct channel_event *chan_ev;
+
+	/* No diffs to register, no journal needed */
+	if (amount_msat_zero(credit_diff)
+	    && amount_msat_zero(debit_diff))
+		return;
+
+	chan_ev = new_channel_event(tmpctx,
+				    tal_fmt(tmpctx, "journal_entry"),
+				    credit_diff,
+				    debit_diff,
+				    AMOUNT_MSAT(0),
+				    currency,
+				    NULL, 0,
+				    timestamp);
+	db_begin_transaction(db);
+	log_channel_event(db, acct, chan_ev);
+	db_commit_transaction(db);
```

Comment:
General: it's often easier and even faster to just wrap the entire event handling in a single large DB transaction rather than trying to do fine-grained control, which may lead to accidentally nested transactions (-> abort), and many transactions being started and committed. Also the one-to-one mapping between event and DB transaction ensure that we don't get partially handled events in case of a crash.